mm2src/coins/utxo/rpc_clients/electrum_rpc/connection_manager/manager.rs

```diff
@@ -0,0 +1,521 @@
+    /// manager without sending them to `utxo_balance_events`. However, we don't do that so that `utxo_balance_events`
+    /// knows about all the added addresses. If it doesn't know about them, it won't be able to retrieve the triggered
+    /// address when its script hash is notified.
+    pub async fn add_subscriptions(&self, addresses: &HashMap<String, Address>) {
+        for (scripthash, address) in addresses.iter() {
+            // For a single address/scripthash, keep trying to subscribe it until we succeed.
+            'single_address_sub: loop {
+                let client = unwrap_or_return!(self.get_client());
+                let connections = self.get_active_connections().await;
+                if connections.is_empty() {
+                    // If there are no active connections, wait for a connection to be established.
+                    Timer::sleep(1.).await;
+                    continue;
+                }
+                // Try to subscribe the address to any connection we have.
+                for connection in connections {
+                    if client
+                        .blockchain_scripthash_subscribe_using(connection.address(), scripthash.clone())
+                        .compat()
+                        .await
+                        .is_ok()
+                    {
+                        let all_connections = self.read_connections();
+                        let connection_ctx = unwrap_or_continue!(all_connections.get(connection.address()));
+                        connection_ctx.add_sub(address.clone());
+                        break 'single_address_sub;
+                    }
+                }
+            }
+        }
+    }
+
+    // Handles the connection event.
+    pub fn on_connected(&self, server_address: &str) {
+        let all_connections = self.read_connections();
+        let connection_ctx = unwrap_or_return!(all_connections.get(server_address));
+
+        // Reset the suspend time & disconnection time.
+        connection_ctx.connected();
+    }
+
+    // Handles the disconnection event from an Electrum server.
+    pub fn on_disconnected(&self, server_address: &str) {
+        let all_connections = self.read_connections();
+        let connection_ctx = unwrap_or_return!(all_connections.get(server_address));
+
+        if self.read_maintained_connections().contains_key(&connection_ctx.id) {
+            // If the connection was maintained, remove it from the maintained connections.
+            let mut maintained_connections = self.write_maintained_connections();
+            maintained_connections.remove(&connection_ctx.id);
+            // And notify the background task if we fell below the `min_connected` threshold.
+            if (maintained_connections.len() as u32) < self.config().min_connected {
+                self.notify_below_min_connected()
+            }
```

Comment:
How about wrapping this up inside a remove connection function? We can do the same for the below https://github.com/KomodoPlatform/komodo-defi-framework/blob/93d79d68f030ff98df6b92e0331b05e8d1b1556d/mm2src/coins/utxo/rpc_clients/electrum_rpc/connection_manager/manager.rs#L417-L422 but it will be add a connection function. We can then remove `write_maintained_connections` function so there is no way for a dev to add or remove a connection without these checks.